When indexed-repeat is used within a repeat, 1st, 2nd, 4th, and 6th a…

gushil opened the pull request:
…rgument not using relative path.

Closes #
Fix #484 
#### Why is this the best possible solution? Were any other approaches considered?
The one that works and passed tests.
#### What are the regression risks?
All tests passed, including new ones.
#### Does this change require updates to documentation? If so, please file an issue [here](https://github.com/XLSForm/xlsform.github.io) and include the link below.
No
#### Before submitting this PR, please make sure you have:
- [x] included test cases for core behavior and edge cases in `tests_v1`
- [x] run `nosetests` and verified all tests pass
- [x] run `black pyxform` to format code
- [x] verified that any code or assets from external sources are properly credited in comments

MartijnR reviewed (commented):
Thanks @gushil!

I checked the tests and this all looks good. I also double-checked the output when indexed-repeat is used in a constraint, relevant and required, and this is also correct.

I glanced at the code a little bit, and all the if-statements seemed unexpected to me. However, I have no understanding of the intricacies of how the current ${node} references are process when referring to nodes inside repeats from that repeat. So, I am hoping this makes immediate sense to @lognaturel...

lognaturel commented:
> all the if-statements seemed unexpected to me.

Agreed, this looks tough to maintain and evolve. Perhaps it's the only reasonable solution but it would really help to have a narrative on what alternatives you considered and why you think this is the best/only approach. Otherwise the only way I can provide meaningful review is to solve the problem myself all over again. It appears that you are explicitly identifying all the different places that an `indexed-repeat` function could appear. Is there not a single place where that could be performed regardless of where `indexed-repeat` appeared? At the very least, this logic should be grouped in a function with a clear comment on its purpose.

gushil commented:
@lognaturel @MartijnR 
Thanks for the reviews!

I agree with your comment about the code looks quite difficult to understand. I try to not moving the code too much, but it looks like it made it difficult to understand and too many conditional repetitions.

Basically, I want to add conditions related to this issue in addition to the existing relative path condition:

`
not (
    context["type"] == "calculate"
    and "indexed-repeat" in context["bind"]["calculate"]
)
`

But it looks like, some of the conditions related to this issue are covered in the existing condition, so I break the current condition and add conditions related to this issue.

Another problem I found is that I cannot tell if this name to be replaced is in indexed-repeat without checking all conditions (context["bind"]["calculate"] and context["default"]), so I made those checks.

I will try to make it simpler and add some comments to it.
I'm open to any suggestions.

Thanks

gushil commented:
Hi, @lognaturel @MartijnR 

Code has been refactored and all additional tests passed.

Hopefully, the current implementation is simpler and easy to understand. I've added some comments too.

Thanks.

MartijnR commented on pyxform/tests_v1/test_repeat.py:
```diff
@@ -458,3 +458,133 @@ def test_choice_from_previous_repeat_answers_in_nested_repeat_uses_current(self)
+            title="dreaded nested repeat indexed-repeat 1st, 2nd, 4th, and 6th parameters is using absolute path",
+            md="""
+                | survey  |                |                |                                  |                                                                                                                 |
+                |         | type           | name           | label                            | default                                                                                                         |
```
Change "default" to "required"? (Though the expression makes little sense in that case - not returning a boolean -, it doesn't really matter for these tests, I think).

gushil replied:
Ok. Let's try using "required". WIll be changed in the next commit.

lognaturel commented:
Thanks for the good rework, @gushil. I'm comfortable with it now.

One case I've found that isn't handled is if the `indexed-repeat` call is nested in another call and there's a relative reference parameter after. For example, consider `concat(indexed-repeat(${name}, ${family}, 1, ${person}, 2), ${name})` in a nested repeat. The output should be `concat(indexed-repeat( /data/family/person/name ,  /data/family , 1,  /data/family/person , 2), ../name )` but is `concat(indexed-repeat( /data/family/person/name ,  /data/family , 1,  /data/family/person , 2),  /data/family/person/name )`.

Putting the `${name}` as the first parameter works: `concat(${name}, indexed-repeat(${name}, ${family}, 1, ${person}, 2))` correctly produces `concat( ../name , indexed-repeat( /data/family/person/name ,  /data/family , 1,  /data/family/person , 2))`.

gushil commented on pyxform/tests_v1/test_repeat.py:
```diff
@@ -458,3 +458,133 @@ def test_choice_from_previous_repeat_answers_in_nested_repeat_uses_current(self)
+        """Test relative path exception (absolute path) in indexed-repeat() using dreaded nested repeat."""
+        self.assertPyxformXform(
+            name="data",
+            title="dreaded nested repeat indexed-repeat 1st, 2nd, 4th, and 6th parameters is using absolute path",
```
Will be removed in next commit

gushil commented on pyxform/tests_v1/test_repeat.py:
```diff
@@ -458,3 +458,133 @@ def test_choice_from_previous_repeat_answers_in_nested_repeat_uses_current(self)
+            title="dreaded nested repeat indexed-repeat 1st, 2nd, 4th, and 6th parameters is using absolute path",
+            md="""
+                | survey  |                |                |                                  |                                                             |
+                |         | type           | name           | label                            | default                                                     |
```
Thanks. Will be fixed in next commit.

gushil commented on pyxform/tests_v1/test_repeat.py:
```diff
@@ -458,3 +458,133 @@ def test_choice_from_previous_repeat_answers_in_nested_repeat_uses_current(self)
+            title="dreaded nested repeat indexed-repeat 1st, 2nd, 4th, and 6th parameters is using absolute path",
+            md="""
+                | survey  |                |                |                                  |                                                        |
+                |         | type           | name           | label                            | default                                                |
```
Got it. WIll be fixed in the next commit.

gushil commented:
Hi @lognaturel @MartijnR 

Addressed @MartijnR reviews and fixed the code (and add a test too) to fix the bug that @lognaturel  mentioned in the feedback.

Thanks.

lognaturel reviewed (approved):
👍 

In general, doing argument parsing with regex (as opposed to a parser) scares me a bit but based on the tests, I think you've pulled it off. I think there are probably still cases like commas in quotes that could break but I think it's ok to wait and see whether it's something users actually end up needing.
